src/main/java/com/google/sps/servlets/AddSaleCardServlet.java

```diff
@@ -99,44 +107,73 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+      // If the vendor doesn't have a saleCard, create completely new Entities (datastore.put(...) to generate Key)
+      // Else, use the already existing Entities
+      if (vendorObject.getSaleCard() == null) {
+        saleCard = new Entity("SaleCard", vendorKey);
+        datastore.put(saleCard); 
+
+        picture = new Entity("Picture", saleCard.getKey());
+        locationData = new Entity("LocationData", saleCard.getKey());
+      } else {
+        saleCard = new Entity("SaleCard",
+            vendorObject.getSaleCard().getId(), vendorKey);
+        picture = new Entity("Picture",
+            vendorObject.getSaleCard().getPicture().getId(), saleCard.getKey());
+        locationData = new Entity("LocationData",
+            vendorObject.getSaleCard().getLocation().getId(), saleCard.getKey());
+      }
 
-      Entity picture = new Entity("Picture", saleCard.getKey());
+      // If not the same, delete previous blob from blobstore
+      if (vendorObject.getSaleCard() != null && 
+          imageBlobKey.compareTo(vendorObject.getSaleCard().getPicture().getBlobKey()) != 0) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        blobstoreService.delete(vendorObject.getSaleCard().getPicture().getBlobKey());
+      }
       picture.setProperty("blobKey", imageBlobKey);
       picture.setProperty("altText", altText);
       datastore.put(picture);
 
-      Entity locationData = new Entity("LocationData", saleCard.getKey());
       locationData.setProperty("salePoint", vendorLocation);
       locationData.setProperty("geoHash", geoHash);
       locationData.setProperty("radius", radius);
       datastore.put(locationData);
 
       EmbeddedEntity picInfo = new EmbeddedEntity();
+      picInfo.setKey(picture.getKey());
       picInfo.setPropertiesFrom(picture);
       EmbeddedEntity locInfo = new EmbeddedEntity();
+      locInfo.setKey(locationData.getKey());
       locInfo.setPropertiesFrom(locationData);
 
+      saleCard.setProperty("businessName", businessName);
+      saleCard.setProperty("description", description);
+      saleCard.setProperty("hasDelivery", hasDelivery);
+      saleCard.setProperty("startTime", startTime);
+      saleCard.setProperty("endTime", endTime);
       saleCard.setIndexedProperty("picture", picInfo);
       saleCard.setIndexedProperty("location", locInfo);
       datastore.put(saleCard);
 
       EmbeddedEntity saleInfo = new EmbeddedEntity();
+      saleInfo.setKey(saleCard.getKey());
       saleInfo.setPropertiesFrom(saleCard);
 
-      /* TODO: Vendor Entity Put
-        Entity vendor = datastore.get(vendorKey);
-        vendor.setIndexedProperty("saleCard", saleInfo);
-        datastore.put(vendor);
-      */
-
+      vendorEntity.setIndexedProperty("saleCard", saleInfo);
+      datastore.put(vendorEntity);
+      
       response.sendRedirect("/");
     } else {
       System.out.println("User is not logged in.");
```

Comment:
You should also send a response to the front end stating that the user must be logged in to use this feature.

Reply:
Hmm... okay, I am already sending an error code, but I'll add a message too!
> The server defaults to creating the response to look like an HTML-formatted server error page containing the specified message, setting the content type to "text/html"